File: Windows/Prosary.Tests/StationsEngineTests.cs
using Prosary.Models;
using Prosary.Services;
using Xunit;

namespace Prosary.Tests;

/// <summary>Mirrors iOS's StationsEngineTests.swift / Android's StationsEngineTest.kt.</summary>
public class StationsEngineTests
{
    private readonly PrayerEngine _engine = new(new LiturgicalCalendarService());

    private IReadOnlyList<RosaryStep> BuildSteps(string? languageCode) =>
        _engine.BuildSteps(new Prayer { Kind = PrayerKind.StationsOfTheCross, LanguageCode = languageCode ?? LanguageCatalog.DefaultSentinel });

    [Fact]
    public void BuildSteps_OpeningPrayerThenFourteenStationsThenClosingPrayer()
    {
        var steps = BuildSteps("en");
        Assert.Equal("Sign of the Cross", steps[0].Title);
        Assert.Equal("Opening Prayer", steps[1].Title);
        Assert.Equal("Closing Prayer", steps[^1].Title);

        var stationSteps = steps.Skip(2).SkipLast(1).ToList();
        Assert.Equal(14, stationSteps.Count);
    }

    [Fact]
    public void BuildSteps_StationsAreInOrderWithCorrectOrdinalSubtitles()
    {
        var stationSteps = BuildSteps("en").Skip(2).SkipLast(1).ToList();
        Assert.Equal("1st Station", stationSteps[0].Subtitle);
        Assert.Equal("14th Station", stationSteps[13].Subtitle);
    }

    [Fact]
    public void BuildSteps_NoStepHasABeadTrackShape()
    {
        // Stations has no decades/beads — the flow UI shows a plain progress bar instead (see
        // ARCHITECTURE.md's "Bead progress track" section).
        var steps = BuildSteps("en");
        Assert.All(steps, s => Assert.True(s.Mystery is null && s.DecadeIndex is null && s.HailMaryIndexInDecade is null));
    }

    [Fact]
    public void BuildSteps_EachStationBodyContainsTheSharedVersicleAndResponse()
    {
        var stationSteps = BuildSteps("en").Skip(2).SkipLast(1);
        Assert.All(stationSteps, s =>
        {
            Assert.Contains("We adore You, O Christ, and we bless You", s.Body);
            Assert.Contains("Because by Your holy Cross You have redeemed the world", s.Body);
        });
    }

    [Fact]
    public void BuildSteps_FirstStationIsCondemnedToDeath()
    {
        var firstStation = BuildSteps("en").Skip(2).First();
        Assert.Equal("Jesus is Condemned to Death", firstStation.Title);
        Assert.Equal("station_01_condemned_to_death", firstStation.ImageOverrideKey);
    }

    [Fact]
    public void BuildSteps_EnglishBodyContainsEnglishText()
    {
        Assert.Contains(BuildSteps("en"), s => s.Body.Contains("We adore You"));
    }

    [Fact]
    public void BuildSteps_LatinBodyContainsLatinText()
    {
        Assert.Contains(BuildSteps("la"), s => s.Body.Contains("Adoramus te, Christe"));
    }
}

File: iOS/ProsaryTests/StationsEngineTests.swift
//
//  StationsEngineTests.swift
//  ProsaryTests
//

import XCTest
@testable import Prosary

final class StationsEngineTests: XCTestCase {
  private func engine() -> PrayerEngine {
    PrayerEngine(calendar: StubLiturgicalCalendar())
  }

  private func steps(languageCode: String = "en") -> [RosaryStep] {
    engine().buildSteps(for: Prayer(kind: .stationsOfTheCross, languageCode: languageCode))
  }

  func testOpeningPrayerThenFourteenStationsThenClosingPrayer() {
    let all = steps()
    XCTAssertEqual(all.first?.title, "Sign of the Cross")
    XCTAssertEqual(all[1].title, "Opening Prayer")
    XCTAssertEqual(all.last?.title, "Closing Prayer")

    let stationSteps = all.dropFirst(2).dropLast()
    XCTAssertEqual(stationSteps.count, 14)
  }

  func testStationsAreInOrderWithCorrectOrdinalSubtitles() {
    let stationSteps = Array(steps().dropFirst(2).dropLast())
    XCTAssertEqual(stationSteps.first?.subtitle, "1st Station")
    XCTAssertEqual(stationSteps[13].subtitle, "14th Station")
  }

  func testNoStepHasABeadTrackShape() {
    // Stations has no decades/beads — the flow UI shows a plain progress bar instead (see
    // ARCHITECTURE.md's "Bead progress track" section).
    let all = steps()
    XCTAssertTrue(all.allSatisfy { $0.mystery == nil && $0.decadeIndex == nil && $0.hailMaryIndexInDecade == nil })
  }

  func testEachStationBodyContainsTheSharedVersicleAndResponse() {
    let stationSteps = Array(steps().dropFirst(2).dropLast())
    for step in stationSteps {
      XCTAssertTrue(step.body.contains("We adore You, O Christ, and we bless You"))
      XCTAssertTrue(step.body.contains("Because by Your holy Cross You have redeemed the world"))
    }
  }

  func testFirstStationIsCondemnedToDeath() {
    let firstStation = steps().dropFirst(2).first
    XCTAssertEqual(firstStation?.title, "Jesus is Condemned to Death")
    XCTAssertEqual(firstStation?.imageKey, "station_01_condemned_to_death")
  }

  func testEnglishBodyContainsEnglishText() {
    XCTAssertTrue(steps(languageCode: "en").contains { $0.body.contains("We adore You") })
  }

  func testLatinBodyContainsLatinText() {
    XCTAssertTrue(steps(languageCode: "la").contains { $0.body.contains("Adoramus te, Christe") })
  }
}
